File: chazm-api/src/main/java/runtimemodels/chazm/api/entity/Agent.java
package runtimemodels.chazm.api.entity;

import runtimemodels.chazm.api.Organization;
import runtimemodels.chazm.api.id.Identifiable;

import java.util.Map;

/**
 * The {@linkplain Agent} interface defines the agent entity of an {@linkplain Organization}.
 *
 * @author Christopher Zhong
 * @since 3.4
 */
public interface Agent extends Identifiable<Agent> {
    /**
     * Returns a {@linkplain Map} with information for contacting this {@linkplain Agent}.
     *
     * @return a {@linkplain Map} with information for contacting this {@linkplain Agent}.
     */
    Map<Object, Object> getContactInfo();

    /**
     * Adds the given {@linkplain Object} {@code key} and {@linkplain Object} {@code value} as a contact information for this {@linkplain Agent}.
     *
     * @param key   an {@linkplain Object} representing the {@code key}.
     * @param value an {@linkplain Object} representing the {@code value}.
     */
    void addContactInfo(Object key, Object value);

    /**
     * Removes a contact information by the given {@linkplain Object} {@code key} from this {@linkplain Agent}.
     *
     * @param key   an {@linkplain Object} representing the {@code key}.
     */
    void removeContactInfo(Object key);
}

File: chazm-api/src/main/kotlin/runtimemodels/chazm/api/entity/Agent.kt
package runtimemodels.chazm.api.entity

import runtimemodels.chazm.api.id.AgentId
import runtimemodels.chazm.api.id.Identifiable
import runtimemodels.chazm.api.organization.Organization

/**
 * The [Agent] interface defines an agent entity of an [Organization].
 *
 * @author Christopher Zhong
 * @since 3.4
 */
interface Agent : Identifiable<Agent, AgentId> {
    /**
     * The [ContactInfo] interface defines a wrapper for a key-value [Map] of [Any].
     *
     * @author Christopher Zhong
     * @since 7.0
     */
    interface ContactInfo : Map<Any, Any> {
        /**
         * Adds the given `key` and `value` as a contact information for this [Agent].
         *
         * @param key   an [Any] representing the `key`.
         * @param value an [Any] representing the `value`.
         */
        fun add(key: Any, value: Any)

        /**
         * Removes a contact information by the given `key` from this [Agent].
         *
         * @param key   an [Any] representing the `key`.
         */
        fun remove(key: Any)
    }

    /**
     * A [ContactInfo] with information for contacting this [Agent].
     */
    val contactInfo: ContactInfo
}
